### PR TITLE
Add `JoinSet`

### PR DESCRIPTION
**Pull request by Darksonn**
This PR adds a new `TaskSet` type that uses Tokio's `linked_list` utility to implement polling in a way that runs in constant time.

Closes: #3903

### CODE REVIEW

File: tokio/src/util/idle_notified_set.rs [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+/// waker and the collection will keep track of whether that waker has been
+/// notified.
+pub(crate) struct IdleNotifiedEntry<T> {
+    entry: Arc<ListEntry<T>>,
+}
+
+struct Lists<T> {
+    inner: Mutex<ListsInner<T>>,
+    /// Only accessed by `IdleNotifiedSet`.
+    length: UnsafeCell<usize>,
```
Comment (carllerche) [inline tokio/src/util/idle_notified_set.rs:32]: Why is this here and not on `IdleNotifiedEntry`?

File: tokio/src/task/task_set.rs [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+                Some(Poll::Ready(Ok(res))) => {
+                    self.inner.with_entry_value(&entry, |jh| jh.take());
+                    self.inner.remove(&entry);
+                    return Poll::Ready(Ok(Some(res)));
+                }
+                Some(Poll::Ready(Err(err))) => {
+                    self.inner.with_entry_value(&entry, |jh| jh.take());
+                    self.inner.remove(&entry);
+                    return Poll::Ready(Err(err));
+                }
```
Comment (carllerche) [inline tokio/src/task/task_set.rs:187]: Looks like the code could be deduped here. I don't *think* we need to re-map it to `Result<Option<T>, ...>` `Option<Result<...>>` should be fine?

File: tokio/src/task/task_set.rs [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    /// Wait until one of the tasks in the set completes and returns its output.
+    ///
+    /// Returns `None` if the set is empty.
+    ///
+    /// # Cancel Safety
+    ///
+    /// This method is cancel safe. If `join_one` is used as the event in a `tokio::select!`
+    /// statement and some other branch completes first, it is guaranteed that no tasks were
+    /// removed from this `TaskSet`.
+    pub async fn join_one(&mut self) -> Result<Option<T>, JoinError> {
```
Comment (carllerche) [inline tokio/src/task/task_set.rs:132]: I would probably keep this as `Option<Result<...>>`.

Comment (Darksonn) [reply tokio/src/task/task_set.rs:132]: Well, it's a pain to use in a `while let` loop if you do that though.

**Review by carllerche — COMMENTED** [review]
The impl looks fine to me. I had a minor API suggestion. I would also suggest we initially release this as an unstable API. We probably should document unstable APIs better though.

**Comment by carllerche** [comment]
@Darksonn How would we add an API in the future for spawning tasks that *don't* abort when the set is dropped?

**Comment by Noah-Kennedy** [comment]
Considering that task set is going to be used a lot for resource cleanup and management, we probably want something along the lines of a `fn shutdown(self) -> impl Future<()>` that aborts all of the tasks and waits for them to finish, so that users can await the termination of the tasks on the set.

**Comment by Darksonn** [comment]
How should we handle panics in the sub-tasks when the user has called `shutdown`?

**Comment by Noah-Kennedy** [comment]
I'm in agreement with Carl here. It would be best not to propagate the panics.

File: tokio/src/task/task_set.rs [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    /// # Return value
+    ///
+    /// This function returns:
+    ///
+    ///  * `Poll::Pending` if the `TaskSet` is not empty and all tasks are currently running.
+    ///  * `Poll::Ready(Some(Ok(value)))` if one of the tasks in this `TaskSet` has completed. The
+    ///    `value` is the return value of one of the tasks that completed.
+    ///  * `Poll::Ready(Some(Err(err)))` if one of the tasks in this `TaskSet` has panicked or been
+    ///     aborted.
+    ///  * `Poll::Ready(None)` if the `TaskSet` is empty.
```
Comment (MattiasBuelens) [inline tokio/src/task/task_set.rs:155]: These don't agree with the return type `Poll<Result<Option<...>>>`.
```suggestion
    ///  * `Poll::Ready(Ok(Some(value)))` if one of the tasks in this `TaskSet` has completed. The
    ///    `value` is the return value of one of the tasks that completed.
    ///  * `Poll::Ready(Err(err))` if one of the tasks in this `TaskSet` has panicked or been
    ///     aborted.
    ///  * `Poll::Ready(Ok(None))` if the `TaskSet` is empty.
```

**Comment by Darksonn** [comment]
I have reworked the `IdleNotifiedSet` to guarantee that the value stored in an entry is dropped when the entry is removed from the set. If a value isn't destroyed when the entry is dropped, then that's a ref-cycle because the entry holds a `JoinHandle` to the task, but the task holds a waker to the entry.

**Comment by carllerche** [comment]
cc @seanmonstar @hawkw

File: tokio/src/task/task_set.rs (hunk 3 above)
Comment (seanmonstar) [reply tokio/src/task/task_set.rs:132]: Is it worse? It seems like returning `Option` means you can loop until all tasks are finished, while as `Result` makes it easier for one task error to blow up the loop.

```rust
while let Some(result) = set.join_one().await {

}
// or
while let Ok(Some(val)) = set.join_one().await {
    // loop canceled on first `Err`?
}
```

The `Option` _does_ make it more annoying if you want to use `?` (like `set.join_one().await?`) in the condition. I have an opinion otherwise about which is better or worse.

File: tokio/src/util/idle_notified_set.rs [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+/// `Neither`, then the `pointers` field may be accessed by the
+/// `IdleNotifiedSet` (this happens inside `drain`).
+///
+/// The `value` field is owned by the `IdleNotifiedSet` and may only be accessed
+/// by the `IdleNotifiedSet`. The operation that sets the value of `my_list` to
+/// `Neither` assumes ownership of the `value`, and it must either drop it or
+/// move it out from this entry to prevent it from getting leaked. (Since the
+/// two linked lists are emptied in the destructor of `IdleNotifiedSet`, the
+/// value should not be leaked.)
+#[repr(C)]
```
Comment (seanmonstar) [inline tokio/src/util/idle_notified_set.rs:91]: Why does this need to be `repr(C)`?

**Review by seanmonstar — APPROVED** [review]

File: tokio/src/util/idle_notified_set.rs (hunk 5 above)
Comment (Darksonn) [reply tokio/src/util/idle_notified_set.rs:91]:
```Rust
unsafe impl<T> linked_list::Link for ListEntry<T> {
    ...

    unsafe fn pointers(
        target: NonNull<ListEntry<T>>,
    ) -> NonNull<linked_list::Pointers<ListEntry<T>>> {
        // Safety: The pointers struct is the first field and ListEntry is
        // `#[repr(C)]` so this cast is safe.
        //
        // We do this rather than doing a field access since `std::ptr::addr_of`
        // is too new for our MSRV.
        target.cast()
    }
}
```

File: tokio/src/task/task_set.rs (hunk 3 above)
Comment (Darksonn) [reply tokio/src/task/task_set.rs:132]: I was mostly thinking of the `while let Some(result) = set.join_one().await?` use-case.

File: tokio/src/task/task_set.rs [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+    /// This method is cancel safe. If `join_one` is used as the event in a `tokio::select!`
+    /// statement and some other branch completes first, it is guaranteed that no tasks were
+    /// removed from this `TaskSet`.
+    pub async fn join_one(&mut self) -> Result<Option<T>, JoinError> {
+        crate::future::poll_fn(|cx| self.poll_join_one(cx)).await
+    }
+
+    /// Abort all tasks on this `TaskSet`.
+    ///
+    /// This does not remove the tasks from the `TaskSet`. To wait for the tasks to complete
```
Comment (Noah-Kennedy) [inline tokio/src/task/task_set.rs:138]: I do feel that it would be more ergonomic for this to abort all tasks _and_ await there completion, but I would be alright with this design as well.

Comment (Noah-Kennedy) [reply tokio/src/task/task_set.rs:138]: We could have a shutdown function which first aborts all tasks and then returns a future which completes when all tasks have completed, and ignores their outputs, including ignoring any panics within the tasks.

File: tokio/src/task/task_set.rs [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+///     while let Some(res) = set.join_one().await.unwrap() {
+///         seen[res] = true;
+///     }
+///
+///     for i in 0..10 {
+///         assert!(seen[i]);
+///     }
+/// }
+/// ```
+pub struct TaskSet<T> {
```
Comment (carllerche) [inline tokio/src/task/task_set.rs:42]: IIRC we discussed renaming this to `JoinSet`? What was the conclusion there? If I remember, the question was whether `TaskSet` would be vague compared to `LocalSet` which is also a "task set" (we have a lot of sets...).

File: tokio/src/task/task_set.rs (hunk 6 above)
Comment (carllerche) [reply tokio/src/task/task_set.rs:138]: Well, right now, you would abort then call `join_one` until drained. There definitely should be a `!async` method for aborting. There could be one that both aborts & waits though.

File: tokio/src/task/task_set.rs [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+    ///     available right now.
+    ///  * `Poll::Ready(Ok(Some(value)))` if one of the tasks in this `TaskSet` has completed. The
+    ///    `value` is the return value of one of the tasks that completed.
+    ///  * `Poll::Ready(Err(err))` if one of the tasks in this `TaskSet` has panicked or been
+    ///     aborted.
+    ///  * `Poll::Ready(Ok(None))` if the `TaskSet` is empty.
+    ///
+    /// Note that this method may return `Poll::Pending` even if one of the tasks has completed.
+    /// This can happen if the coop budget is reached.
+    pub fn poll_join_one(&mut self, cx: &mut Context<'_>) -> Poll<Result<Option<T>, JoinError>> {
```
Comment (carllerche) [inline tokio/src/task/task_set.rs:176]: I would avoid making this public for now. It *might* make it trickier to implement `AsyncIterator` in the future. I have not thought through this yet though.

File: tokio/src/task/task_set.rs (hunk 7 above)
Comment (Noah-Kennedy) [reply tokio/src/task/task_set.rs:42]: I prefer TaskSet because it goes along with the broader scoping/cancellation features.

File: tokio/src/task/task_set.rs (hunk 6 above)
Comment (Noah-Kennedy) [reply tokio/src/task/task_set.rs:138]: I would agree, it just might be nice to have this _and_ a more convenient method to wait for all to complete after the mass abort.

**Review by carllerche — COMMENTED** [review]
Looking good. I left some comments inline.

Question: Do we want to start with making this an unstable API or do we want to go straight to stable?

File: tokio/src/task/task_set.rs (hunk 3 above)
Comment (carllerche) [reply tokio/src/task/task_set.rs:132]: I think I agree w/ @Darksonn but what is the precedence in `std`? I'm not sure what the best way to search is. I can't think of any methods that returns `Result<Option<_>, _>` and the only methods that return `Option<Result<...>>` would be `Iterator`, which doesn't really count.

**Review by hawkw — COMMENTED** [review]
haven't looked at the actually interesting code in `idle_notified_set.rs` yet, but here's some misc thoughts

File: tokio/src/task/task_set.rs (hunk 3 above)
Comment (Darksonn) [reply tokio/src/task/task_set.rs:132]: I do not think that std has precedence on this.

File: tokio/src/task/task_set.rs [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+    /// a loop until it returns `Ok(None)`.
+    ///
+    /// This method ignores any panics in the tasks shutting down. When this call returns, the
+    /// `JoinSet` will be empty.
+    ///
+    /// [`abort_all`]: fn@Self::abort_all
+    /// [`join_one`]: fn@Self::join_one
+    pub async fn shutdown(&mut self) {
+        self.abort_all();
+        while self.join_one().await.transpose().is_some() {}
```
Comment (Noah-Kennedy) [inline tokio/src/task/task_set.rs:151]: I think it would be possibly advantageous to bifurcate this operation, and have the function return impl future and be non-async. That way, we could abort the tasks, then return a future that awaits all tasks. This does add complexity though.

Comment (Noah-Kennedy) [reply tokio/src/task/task_set.rs:151]: @Darksonn if we made this change in a future PR, would it be considered blocking?

**Comment by Noah-Kennedy** [comment]
This seems about ready. I left a few questions and comments, but I don't think they block a merge here.

**Review by Noah-Kennedy — APPROVED** [review]

File: tokio/src/util/idle_notified_set.rs [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+/// Note: This entry can be moved from the idle to the notified list while this
+/// object exists by waking its waker.
+pub(crate) struct EntryInOneOfTheLists<'a, T> {
+    entry: Arc<ListEntry<T>>,
+    set: &'a mut IdleNotifiedSet<T>,
+}
+
+struct Lists<T> {
+    inner: Mutex<ListsInner<T>>,
+}
```
Comment (hawkw) [inline tokio/src/util/idle_notified_set.rs:45]: nit, take it or leave it: i'm not totally convinced this adds value over something like
```suggestion
type Lists<T> = Arc<Mutex<ListsInner<T>>>;
```
AFAICT, all this does is makes it so that you have to write `.inner` every time you lock the lists?

File: tokio/src/util/idle_notified_set.rs [hunk 11]
```diff
@@ -0,0 +1,463 @@
...
+            value: UnsafeCell::new(ManuallyDrop::new(value)),
+            my_list: UnsafeCell::new(List::Idle),
+            pointers: linked_list::Pointers::new(),
+            _pin: PhantomPinned,
+        });
+
+        {
+            let mut lock = self.lists.inner.lock();
+            lock.idle.push_front(entry.clone());
+        }
```
Comment (hawkw) [inline tokio/src/util/idle_notified_set.rs:162]: nit: can't this just be
```suggestion
        self.lists.inner.lock().idle.push_front(entry.clone());
```

File: tokio/src/util/idle_notified_set.rs [hunk 12]
```diff
@@ -0,0 +1,463 @@
...
+            None => true,
+        };
+        if should_update_waker {
+            lock.waker = Some(waker.clone());
+        }
+
+        // Pop the entry.
+        match lock.notified.pop_back() {
+            Some(entry) => {
+                // Safety: We are holding the lock.
```
Comment (hawkw) [inline tokio/src/util/idle_notified_set.rs:191]: nit: should the safety comment be moved so it's directly above the `unsafe` block?

File: tokio/src/util/idle_notified_set.rs [hunk 13]
```diff
@@ -0,0 +1,463 @@
...
+                entry.my_list.with_mut(|ptr| unsafe {
+                    *ptr = List::Idle;
+                });
+                drop(lock);
+
+                // Safety: We just put the entry in the idle list, so it is in one of the lists.
+                Some(EntryInOneOfTheLists { entry, set: self })
+            }
+            None => None,
+        }
```
Comment (hawkw) [inline tokio/src/util/idle_notified_set.rs:202]: this seems like it could be simplified to:
```suggestion
        // Pop the entry.
        let entry = lock.notified.pop_back()?;
        lock.idle.push_front(entry.clone());
        // Safety: We are holding the lock.
        entry.my_list.with_mut(|ptr| unsafe {
            *ptr = List::Idle;
        });

        // Safety: We just put the entry in the idle list, so it is in one of the lists.
        Some(EntryInOneOfTheLists { entry, set: self })
```

File: tokio/src/util/idle_notified_set.rs [hunk 14]
```diff
@@ -0,0 +1,463 @@
...
+    }
+
+    /// Call a function on every element in this list.
+    pub(crate) fn for_each<F: FnMut(&mut T)>(&mut self, mut func: F) {
+        fn get_ptrs<T>(list: &mut LinkedList<T>, ptrs: &mut Vec<*mut T>) {
+            let mut node = list.last();
+
+            while let Some(entry) = node {
+                ptrs.push(entry.value.with_mut(|ptr| {
+                    let ptr: *mut ManuallyDrop<T> = ptr;
```
Comment (hawkw) [inline tokio/src/util/idle_notified_set.rs:212]: what is this `let` doing? AFAICT, the `with_mut` closure could be simplified to just:
```rust
ptrs.push(entry.value.with_mut(|ptr| {
    ptr.cast::<T>()
}));
```
so if there's a reason it *can't* be, i'd like to know why?

File: tokio/src/util/idle_notified_set.rs [hunk 15]
```diff
@@ -0,0 +1,463 @@
...
+        //
+        // If the closure panics, then the destructor of the `AllEntries` bomb
+        // ensures that we keep running the destructor on the remaining values.
+        // A second panic will abort the program.
+        while all_entries.pop_next() {}
+    }
+}
+
+/// The mutex for the entries must be held, and the target list must be such
+/// that setting `my_list` to `Neither` is ok.
```
Comment (hawkw) [inline tokio/src/util/idle_notified_set.rs:317]: very minor nit: should this be
```suggestion
/// # Safety
///
/// The mutex for the entries must be held, and the target list must be such
/// that setting `my_list` to `Neither` is ok.
```

File: tokio/src/util/idle_notified_set.rs [hunk 16]
```diff
@@ -0,0 +1,463 @@
...
+            let old_my_list = self.entry.my_list.with_mut(|ptr| unsafe {
+                let old_my_list = *ptr;
+                *ptr = List::Neither;
+                old_my_list
+            });
+
+            let list = match old_my_list {
+                List::Idle => &mut lock.idle,
+                List::Notified => &mut lock.notified,
+                List::Neither => unreachable!(),
```
Comment (hawkw) [inline tokio/src/util/idle_notified_set.rs:350]: nitpick, not a big deal: it would be nice if the `unreachable!` macro said why it's unreachable, even though it is admittedly a bit obvious here...

File: tokio/src/util/idle_notified_set.rs [hunk 17]
```diff
@@ -0,0 +1,463 @@
...
+                }));
+
+                let prev = entry.pointers.get_prev();
+                node = prev.map(|prev| unsafe { &*prev.as_ptr() });
+            }
+        }
+
+        // Atomically get a raw pointer to the value of every entry.
+        //
+        // Since this only locks the mutex once, its not possible for a value to
```
Comment (hawkw) [inline tokio/src/util/idle_notified_set.rs:224]: nit, typo:
```suggestion
        // Since this only locks the mutex once, it's not possible for a value to
```

File: tokio/src/util/idle_notified_set.rs [hunk 18]
```diff
@@ -0,0 +1,463 @@
...
+///
+/// The `value` field is owned by the `IdleNotifiedSet` and may only be accessed
+/// by the `IdleNotifiedSet`. The operation that sets the value of `my_list` to
+/// `Neither` assumes ownership of the `value`, and it must either drop it or
+/// move it out from this entry to prevent it from getting leaked. (Since the
+/// two linked lists are emptied in the destructor of `IdleNotifiedSet`, the
+/// value should not be leaked.)
+///
+/// This type is #[repr(C)] because its `linked_list::Link` implementation uses
+/// that `pointers` is the first field.
```
Comment (hawkw) [inline tokio/src/util/idle_notified_set.rs:93]: nit: maybe rephrase this to
```suggestion
/// This type is `#[repr(C)]` because its `linked_list::Link` implementation requires
/// that `pointers` is the first field.
```

File: tokio/src/util/idle_notified_set.rs [hunk 19]
```diff
@@ -0,0 +1,463 @@
...
+    fn drop(&mut self) {
+        // Clear both lists.
+        self.drain(drop);
+
+        #[cfg(debug_assertions)]
+        {
+            let lock = self.lists.inner.lock();
+            assert!(lock.idle.is_empty());
+            assert!(lock.notified.is_empty());
+        }
```
Comment (hawkw) [inline tokio/src/util/idle_notified_set.rs:398]: nit: should this maybe be
```suggestion
        #[cfg(debug_assertions)]
        if !std::thread::panicking() {
            let lock = self.lists.inner.lock();
            assert!(lock.idle.is_empty());
            assert!(lock.notified.is_empty());
        }
```
so that the assertions won't double panic?

File: tokio/src/util/idle_notified_set.rs [hunk 20]
```diff
@@ -0,0 +1,463 @@
+//! This module defines an `IdleNotifiedSet`, which is a collection of elements.
+//! Each element is intended to correspond to a task, and the collection will
+//! keep track of which tasks have notified their waker, and which have not.
```
Comment (hawkw) [inline tokio/src/util/idle_notified_set.rs:3]: nit, take it or leave it: should this maybe be rephrased as 
```suggestion
//! keep track of which tasks have had their wakers notified, and which have not.
```
(it feels like the current phrasing kind of implies that it tracks whether the tasks have woken their _own_ waker, while in actuality, the wakeups may also be external to the task)

File: tokio/src/util/idle_notified_set.rs [hunk 21]
```diff
@@ -0,0 +1,463 @@
...
+    pub(crate) fn drain<F: FnMut(T)>(&mut self, func: F) {
+        if self.length == 0 {
+            // Fast path.
+            return;
+        }
+        self.length = 0;
+
+        // The LinkedList is not cleared on panic, so we use a bomb to clear it.
+        //
+        // This value has the invariant that eny entry in its `all_entries` list
```
Comment (hawkw) [inline tokio/src/util/idle_notified_set.rs:261]: nit, typo:
```suggestion
        // This value has the invariant that any entry in its `all_entries` list
```

File: tokio/src/task/task_set.rs [hunk 22]
```diff
@@ -0,0 +1,242 @@
...
+        let mut entry = match self.inner.pop_notified(cx.waker()) {
+            Some(entry) => entry,
+            None => {
+                if self.is_empty() {
+                    return Poll::Ready(Ok(None));
+                } else {
+                    // The waker was set by `pop_notified`.
+                    return Poll::Pending;
+                }
+            }
```
Comment (hawkw) [inline tokio/src/task/task_set.rs:208]: nit, take it or leave it: IMO, this is slightly nicer, but I don't have a strong opinion here really
```suggestion
            None if self.is_empty() => return Poll::Ready(Ok(None)),
            // The waker was set by `pop_notified`.
            None => return Poll::Pending,
```

**Review by hawkw — COMMENTED** [review]
Okay, I gave the `IdleNotifiedSet` code a closer look, and it seems correct to me. I had some minor, non-blocking style suggestions; feel free to ignore any you disagree with.

File: tokio/src/util/idle_notified_set.rs (hunk 11 above)
Comment (Noah-Kennedy) [reply tokio/src/util/idle_notified_set.rs:162]: I see that I am not the only one afflicted by this lifetime/scoping paranoia.

File: tokio/src/task/task_set.rs (hunk 9 above)
Comment (Darksonn) [reply tokio/src/task/task_set.rs:151]: No, I don't think it would be considered a breaking change.

File: tokio/src/util/idle_notified_set.rs (hunk 14 above)
Comment (Darksonn) [reply tokio/src/util/idle_notified_set.rs:212]: I was mostly just putting in a bunch of type hints to convince myself the pointers had the types I expected.

File: tokio/src/task/task_set.rs [hunk 23]
```diff
@@ -0,0 +1,222 @@
...
+    ///
+    /// # Cancel Safety
+    ///
+    /// This method is cancel safe. If `join_one` is used as the event in a `tokio::select!`
+    /// statement and some other branch completes first, it is guaranteed that no tasks were
+    /// removed from this `TaskSet`.
+    pub async fn join_one(&mut self) -> Result<Option<T>, JoinError> {
+        crate::future::poll_fn(|cx| self.poll_join_one(cx)).await
+    }
+
```
[the remark this answers is not in the document]
Comment (Darksonn) [reply tokio/src/task/task_set.rs]: I think that the details are unclear wrt. returning a `Vec<Result<T, JoinError>>` because it is not `#[must_use]` and I'm somewhat tired of seeing people ignoring errors when using `join_all`.

File: tokio/src/task/task_set.rs (hunk 22 above)
Comment (Darksonn) [reply tokio/src/task/task_set.rs]: This runs into a borrow-check error because `self` is still borrowed by the return value of `pop_notified`.

Comment (hawkw) [reply tokio/src/task/task_set.rs]: ah, yup. nevermind, then!

File: tokio/src/util/idle_notified_set.rs (hunk 14 above)
Comment (hawkw) [reply tokio/src/util/idle_notified_set.rs:207]: ah, that's fine, then!

File: tokio/src/util/idle_notified_set.rs [hunk 24]
```diff
@@ -0,0 +1,461 @@
...
+///
+/// The `value` field is owned by the `IdleNotifiedSet` and may only be accessed
+/// by the `IdleNotifiedSet`. The operation that sets the value of `my_list` to
+/// `Neither` assumes ownership of the `value`, and it must either drop it or
+/// move it out from this entry to prevent it from getting leaked. (Since the
+/// two linked lists are emptied in the destructor of `IdleNotifiedSet`, the
+/// value should not be leaked.)
+///
+/// This type is `#[repr(C)]` because its `linked_list::Link` implementation
+/// requires that `pointers` is the first field.
```
Comment (hawkw) [inline tokio/src/util/idle_notified_set.rs:91]: nit, take it or leave it: should this all be in a `# Safety` heading, like:
```suggestion
/// # Safety
///
/// This type has a number of safety invariants:
///
/// The `my_list` field must only be accessed while holding the mutex in
/// `parent`. It is an invariant that the value of `my_list` corresponds to
/// which linked list in the `parent` holds this entry. Once this field takes
/// the value `Neither`, then it may never be modified again.
///
/// If the value of `my_list` is `Notified` or `Idle`, then the `pointers` field
/// must only be accessed while holding the mutex. If the value of `my_list` is
/// `Neither`, then the `pointers` field may be accessed by the
/// `IdleNotifiedSet` (this happens inside `drain`).
///
/// The `value` field is owned by the `IdleNotifiedSet` and may only be accessed
/// by the `IdleNotifiedSet`. The operation that sets the value of `my_list` to
/// `Neither` assumes ownership of the `value`, and it must either drop it or
/// move it out from this entry to prevent it from getting leaked. (Since the
/// two linked lists are emptied in the destructor of `IdleNotifiedSet`, the
/// value should not be leaked.)
///
/// This type is `#[repr(C)]` because its `linked_list::Link` implementation
/// requires that `pointers` is the first field.
```

**Review by hawkw — APPROVED** [review]
okay, this looks great to me! thanks @Darksonn for addressing my various nitpicks. :)
